File: EnumerateFilesAsync/Classes/Operations.vb
Imports System.IO

Namespace Classes
    Public Class Operations
        Public Event OnIterate As DelegatesModule.OnIterate
        Private folderFound As Boolean
        Public ReadOnly Property FolderExists() As Boolean
            Get
                Return folderFound
            End Get
        End Property
        Public Async Function IterateFolders(folderName As String, searchString As String, token As Threading.CancellationToken) As Task(Of List(Of String))

            Dim foundList = New List(Of String)()

            If Not Directory.Exists(folderName) Then
                OnIterateEvent?.Invoke(New MonitorProgressArgs("Folder does not exists!"))
                folderFound = False
                Return foundList
            End If

            folderFound = True
            Dim currentFileName = ""
            Try
                For Each fileName In Directory.EnumerateFiles(folderName, "*.txt", SearchOption.AllDirectories)
                    currentFileName = fileName
                    OnIterateEvent?.Invoke(New MonitorProgressArgs(fileName))

                    Dim contents = File.ReadAllText(fileName)

                    Await Task.Delay(1)

                    If contents.Contains(searchString) Then
                        foundList.Add(fileName)
                    End If

                    If token.IsCancellationRequested Then
                        token.ThrowIfCancellationRequested()
                    End If

                Next
            Catch ex As Exception
                OnIterateEvent?.Invoke(New MonitorProgressArgs($"Access denied {currentFileName}"))
            End Try

            Return foundList

        End Function
    End Class
End Namespace

File: EnumerateFilesAsyncSharp/Classes/Operations.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace EnumerateFilesAsync.Classes
{
	namespace Classes
	{
		public class Operations
		{
			public event DelegatesModule.OnIterate OnIterate;
			private bool _folderFound;
			public bool FolderExists => _folderFound;

            public async Task<List<string>> IterateFolders(string folderName, string searchString, CancellationToken token)
			{

				var foundList = new List<string>();

				if (!Directory.Exists(folderName))
				{
					OnIterate?.Invoke(new MonitorProgressArgs("Folder does not exists!"));
					_folderFound = false;
					return foundList;
				}

				_folderFound = true;
				var currentFileName = "";

				try
				{
					foreach (var fileName in Directory.EnumerateFiles(folderName, "*.txt", SearchOption.AllDirectories))
					{
						currentFileName = fileName;
						OnIterate?.Invoke(new MonitorProgressArgs(fileName));

						var contents = File.ReadAllText(fileName);

						await Task.Delay(1, token);

						if (contents.Contains(searchString))
						{
							foundList.Add(fileName);
						}

						if (token.IsCancellationRequested)
						{
							token.ThrowIfCancellationRequested();
						}

					}
				}
				catch (Exception ex)
				{
					OnIterate?.Invoke(new MonitorProgressArgs($"Access denied {currentFileName}"));
				}

				return foundList;

			}
		}
	}

}